transport.go

```diff
@@ -150,13 +150,31 @@ func (t *transport) handle() {
 	for {
 		select {
 		case req := <-t.requests:
-			// BUG(stevvooe): This is an awful tag allocation procedure.
-			// Replace this with something that let's us allocate tags and
-			// associate data with them, returning to them to a pool when
-			// complete. Such a system would provide a lot of information
-			// about outstanding requests.
-			tags++
-			fcall := newFcall(tags, req.message)
+			var selected Tag
+			if req.message.Type() == Tversion {
+				// Tversion can only use NOTAG.
+				if _, exists := outstanding[NOTAG]; exists {
+					req.err <- errors.New("NOTAG already in use")
+				}
+				selected = NOTAG
+			} else {
+				// The tag pool is depleted if all 65536 tags are taken, or if
+				// 65535 tags are taken and NOTAG is available.
+				if len(outstanding) > 0xFFFF {
+					req.err <- errors.New("tag pool depleted")
+				} else if len(outstanding) == 0xFFFF {
+					if _, exists := outstanding[NOTAG]; !exists {
+						req.err <- errors.New("tag pool depleted")
+					}
+				}
+
+				exists := true
+				for selected = 0; exists && selected < NOTAG; selected++ {
```

Comment:
> Even with thousand tags open, the overhead of using a permanent array is quite significant.

Go has a feature called slices which can avoid a full allocation of the tag space. It becomes a little more tricky that a map when handling wraparound, but we the memory usage is well-bounded in the common case vs the map, which can allocate a lot of extra memory.

Albeit, I'd be satisfied if we could simply avoid the linear scan through the maps keyspace. A hint plus linear probe should be sufficient.